presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java

```diff
@@ -82,10 +88,12 @@ public SerializedPage serialize(Page page)
         else {
             //  Encryption disabled, slice data is likely much smaller than its backing buffer
             //  either because of compression or dynamic sizing of the initial output slice
-            slice = Slices.copyOf(slice);
+            if (slice.length() < uncompressedSize) {
```

Comment:
I just thought it over and also think it's better not to change the orginal serialize(Page page) method. For OptimizedPartitionedOutputOperator, the passed in slice is precisely sized and shouldn't be copied at here. Copying it incurs 10% overhead overall. Combining your previous suggestions I made the code like follows:
```
public SerializedPage serialize(Slice slice, int positionCount)
    {
        int uncompressedSize = slice.length();
        byte markers = PageCodecMarker.none();

        if (compressor.isPresent()) {
            int maxCompressedSize = compressor.get().maxCompressedLength(uncompressedSize);
            ensureCompressionBufferCapacity(maxCompressedSize);
            int compressedSize = compressor.get().compress((byte[]) slice.getBase(), 0, uncompressedSize, compressionBuffer, 0, maxCompressedSize);

            if ((((double) compressedSize) / uncompressedSize) <= MINIMUM_COMPRESSION_RATIO) {
                slice = Slices.wrappedBuffer(compressionBuffer, 0, compressedSize);
                markers = COMPRESSED.set(markers);
            }
        }

        if (spillCipher.isPresent()) {
            slice = Slices.wrappedBuffer(spillCipher.get().encrypt(slice.toByteBuffer()));
            markers = ENCRYPTED.set(markers);
        }

        // The slice is either compressed but compression rate > MINIMUM_COMPRESSION_RATIO, or not compressed.
        // In either case, we don't want to copy the slice because the slice passed in by OptimizedPartitionedOutputOperator
        // should be precisely sized.
        return new SerializedPage(slice, markers, positionCount, uncompressedSize);
    }

    private void ensureCompressionBufferCapacity(int capacity)
    {
        if (compressionBuffer == null || compressionBuffer.length < capacity) {
            compressionBuffer = new byte[capacity];
        }
    }
```

Reply:
Small nit, when I ported the PR that introduced this logic to PrestoSQL (and modified it in response to review feedback) the logic was changed to check `slice.isCompact()` to determine whether to create a new copy. See: https://github.com/prestosql/presto/pull/778/files#diff-8d88d4025075988a194c581a83234de7R97

Reply:
Thanks @pettyjamesm for sharing the change in prestosql. I can add this isCompact() check but in reality it should not happen, i.e. the uncompressed, unciphered original slice should always be compact, because the OptimizedPartitionedOutputOperator calculate the size before creating the SliceOutput.
```
// The original function is not changed
public SerializedPage serialize(Page page)
    {
        SliceOutput serializationBuffer = new DynamicSliceOutput(toIntExact(page.getSizeInBytes() + Integer.BYTES)); // block length is an int
        writeRawPage(page, serializationBuffer, blockEncodingSerde);
        Slice slice = serializationBuffer.slice();
        int uncompressedSize = serializationBuffer.size();
        byte markers = PageCodecMarker.none();

        if (compressor.isPresent()) {
            ByteBuffer compressionBuffer = ByteBuffer.allocate(compressor.get().maxCompressedLength(uncompressedSize));
            compressor.get().compress(slice.toByteBuffer(), compressionBuffer);
            compressionBuffer.flip();
            if ((((double) compressionBuffer.remaining()) / uncompressedSize) <= MINIMUM_COMPRESSION_RATIO) {
                slice = Slices.wrappedBuffer(compressionBuffer);
                markers = COMPRESSED.set(markers);
            }
        }

        if (spillCipher.isPresent()) {
            slice = Slices.wrappedBuffer(spillCipher.get().encrypt(slice.toByteBuffer()));
            markers = ENCRYPTED.set(markers);
        }
        else {
            //  Encryption disabled, slice data is likely much smaller than its backing buffer
            //  either because of compression or dynamic sizing of the initial output slice
            slice = Slices.copyOf(slice);
        }

        return new SerializedPage(slice, markers, page.getPositionCount(), uncompressedSize);
    }

public SerializedPage serialize(Slice slice, int positionCount)
    {
        int uncompressedSize = slice.length();
        byte markers = PageCodecMarker.none();

        if (compressor.isPresent()) {
            int maxCompressedSize = compressor.get().maxCompressedLength(uncompressedSize);
            ensureCompressionBufferCapacity(maxCompressedSize);
            int compressedSize = compressor.get().compress((byte[]) slice.getBase(), 0, uncompressedSize, compressionBuffer, 0, maxCompressedSize);

            if ((((double) compressedSize) / uncompressedSize) <= MINIMUM_COMPRESSION_RATIO) {
                slice = Slices.copyOf(Slices.wrappedBuffer(compressionBuffer, 0, compressedSize));
                markers = COMPRESSED.set(markers);
            }
        }

        if (spillCipher.isPresent()) {
            slice = Slices.wrappedBuffer(spillCipher.get().encrypt(slice.toByteBuffer()));
            markers = ENCRYPTED.set(markers);
        }

        // The slice is either compressed but compression rate > MINIMUM_COMPRESSION_RATIO, or not compressed.
        // In either case, we don't want to copy the slice because the slice passed in by OptimizedPartitionedOutputOperator
        // should be precisely sized. We added the isCompact() check here but the copy should never happen.
        if (!slice.isCompact()) {
            slice = Slices.copyOf(slice);
        }

        return new SerializedPage(slice, markers, positionCount, uncompressedSize);
    }
```